UNIXFS.md

```diff
@@ -59,6 +60,32 @@ For files that are comprised of more than a single block, the 'Type' field will
 
 This data is serialized and placed inside the 'Data' field of the outer merkledag protobuf, which also contains the actual links to the child nodes of this object.
 
+For files comprised of a single block, the 'Type' field will be set to 'File', 'filesize' will be set to the total number of bytes in the file and the file data will be stored in the 'Data' field.
+
+The serialized size of a UnixFS node must not exceed 256KiB in order to work will with the [Bitswap] protocol.
+
+## Metadata
+
+UnixFS currently supports three optional metadata fields:
+
+* `mode` -- The `mode` is for persisting the [file permissions in numeric notation](https://en.wikipedia.org/wiki/File_system_permissions#Numeric_notation) \[[spec](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_stat.h.html)\]. If unspecified this defaults to `0755` for directories/HAMT shards and `0644` for all other types where applicable.
+* `mtime` -- The modification time in seconds since the epoch. This defaults to the unix epoch if unspecified.
+* `MimeType` -- The mime type of the file. This field is deprecated, as is the `Metadata` message and neither will be present in `UnixFSv2` because mime types are not typically properties of a file system.
+
+### Inheritance
+
+When traversing down through a UnixFSv1.5 directory, child entries without metadata fields will inherit those of their direct ascendants.
+
+### Deduplication and inlining
+
+Where the file data is small it would normally be stored in the `Data` field of the UnixFS `File` node.
+
+To aid in deduplication of data even for small files, file data can be stored in a separate node linked to from the `File` node in order for the data to have a constant [CID] regardless of the metadata associated with it.
+
+As a further optimization, if the `File` node's serialized size is small, it may be inlined into it's v1 [CID] by using the [`identity`](https://github.com/multiformats/multicodec/blob/master/table.csv) [multihash].
+
+Such [CID]s must consist of 23 bytes or fewer in order for them to fit inside the 63 character limit for a DNS label when encoded in base32 (see [RFC1035 Section 2.3.1](https://tools.ietf.org/html/rfc1035#section-2.3.1)).
```

Comment:
* Could you do a quick check to see if this is actually possible?
* In terms of domains, we really need to generalize to multiple subdomains (different issue).

Reply:
I think so, as long as we don't get to the extremes of any number values.  For 'reasonable' values it should work:

```javascript
const CID = require('cids')
const multihashing = require('multihashing-async')
const UnixFS = require('ipfs-unixfs')

async function main () {
  const data = new UnixFS('file')
  data.addBlockSize(256)

  const multihash = await multihashing(data.marshal(), 'identity')
  const cid = new CID(1, 'dag-pb', multihash)

  console.info(cid.toBaseEncodedString('base32').length)
  // prints '21'
}

main()
```

For the following configurations:

```javascript
const data = new UnixFS('file')
data.addBlockSize(256)
data.fanout = 4294967295  // e.g. max uint32

// prints '30'
```

Reply:
Nice!

Reply:
nit: should probably 23 + a single link to a block.

Reply:
I've pulled this out too because the generated identity CID is too long to be used in DNS labels after all:

```javascript
const CID = require('cids')
const multihashing = require('multihashing-async')
const UnixFS = require('ipfs-unixfs')
const dagPB = require('ipld-dag-pb')
const { DAGLink, DAGNode } = dagPB

async function main () {
  // store some file data
  const fileData = Buffer.from([0, 1, 2, 3, 4])
  const fileDataHash = await multihashing(fileData, 'sha2-256')
  const fileDataCid = new CID(1, 'raw', fileDataHash)

  // create the UnixFS node that points to the file data
  const file = new UnixFS('file')
  file.addBlockSize(5)
  const node = new DAGNode(file.marshal(), [
    new DAGLink('', 5, fileDataCid)
  ])

  // inline the UnixFS node into it's CID
  const multihash = await multihashing(dagPB.util.serialize(node), 'identity')
  const cid = new CID(1, 'dag-pb', multihash)

  console.info(cid.toBaseEncodedString('base32').length)
  // prints '91'
}

main()

```